GCS SDK does not resume downloads that fail part-way through

A connection dropped part-way through the download of a blob from GCS is not today considered retryable by the SDK, so a restore from GCS over a flaky connection may fail and need to be retried from the beginning.

I've seen two different (but related) exceptions:

<details><p>

```
[2020-02-13T12:10:57,150][WARN ][o.e.c.r.a.AllocationService] [elasticsearch-2] failing shard [failed shard, shard [elasticlogs-2018-05-13][0], node[d2sAR621RziQbILiZMBvfg], [P], recovery_source[snapshot recovery [iJoJDpgyR9GjOYO1enQmjQ] from small_logging_cluster:snapshot_80gb/fioJzVMyTcKPCRdfN5lZ7w], s[INITIALIZING], a[id=hi-CTSfzRliMbh3ajtzVVg], unassigned_info[[reason=NEW_INDEX_RESTORED], at[2020-02-13T11:22:30.460Z], delayed=false, details[restore_source[small_logging_cluster/snapshot_80gb]], allocation_status[deciders_throttled]], message [failed recovery], failure [RecoveryFailedException[[elasticlogs-2018-05-13][0]: Recovery failed on {elasticsearch-1}{d2sAR621RziQbILiZMBvfg}{cgzSxj1tSjawYThZfSEXpQ}{10.10.9.72}{10.10.9.72:9300}{dim}{xpack.installed=true}]; nested: IndexShardRecoveryException[failed recovery]; nested: IndexShardRestoreFailedException[restore failed]; nested: IndexShardRestoreFailedException[failed to restore snapshot [snapshot_80gb/fioJzVMyTcKPCRdfN5lZ7w]]; nested: NotSerializableExceptionWrapper[storage_exception: Connection reset]; nested: IOException[Connection reset]; nested: IOException[Connection reset]; ], markAsStale [true]]
org.elasticsearch.indices.recovery.RecoveryFailedException: [elasticlogs-2018-05-13][0]: Recovery failed on {elasticsearch-1}{d2sAR621RziQbILiZMBvfg}{cgzSxj1tSjawYThZfSEXpQ}{10.10.9.72}{10.10.9.72:9300}{dim}{xpack.installed=true}
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2584) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:703) ~[elasticsearch-7.5.1.jar:7.5.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:830) [?:?]
Caused by: org.elasticsearch.index.shard.IndexShardRecoveryException: failed recovery
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:353) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:283) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1867) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2580) ~[elasticsearch-7.5.1.jar:7.5.1]
	... 4 more
Caused by: org.elasticsearch.index.snapshots.IndexShardRestoreFailedException: restore failed
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:480) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromRepository$5(StoreRecovery.java:285) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:308) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:283) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1867) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2580) ~[elasticsearch-7.5.1.jar:7.5.1]
	... 4 more
Caused by: org.elasticsearch.index.snapshots.IndexShardRestoreFailedException: failed to restore snapshot [snapshot_80gb/fioJzVMyTcKPCRdfN5lZ7w]
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.restoreShard(BlobStoreRepository.java:1255) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.Repository.restoreShard(Repository.java:231) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:469) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromRepository$5(StoreRecovery.java:285) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:308) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:283) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1867) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2580) ~[elasticsearch-7.5.1.jar:7.5.1]
	... 4 more
Caused by: org.elasticsearch.common.io.stream.NotSerializableExceptionWrapper: storage_exception: Connection reset
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:227) ~[?:?]
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.read(HttpStorageRpc.java:690) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:127) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:124) ~[?:?]
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105) ~[?:?]
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76) ~[?:?]
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel.read(BlobReadChannel.java:123) ~[?:?]
	at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.lambda$read$0(GoogleCloudStorageBlobStore.java:189) ~[?:?]
	at java.security.AccessController.doPrivileged(AccessController.java:554) ~[?:?]
	at org.elasticsearch.repositories.gcs.SocketAccess.doPrivilegedIOException(SocketAccess.java:44) ~[?:?]
	at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.read(GoogleCloudStorageBlobStore.java:189) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:65) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:109) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:103) ~[?:?]
	at org.elasticsearch.index.snapshots.blobstore.SlicedInputStream.read(SlicedInputStream.java:92) ~[elasticsearch-7.5.1.jar:7.5.1]
	at java.io.InputStream.read(InputStream.java:213) ~[?:?]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFile(FileRestoreContext.java:240) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFiles(FileRestoreContext.java:217) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restore(FileRestoreContext.java:180) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.restoreShard(BlobStoreRepository.java:1253) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.Repository.restoreShard(Repository.java:231) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:469) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromRepository$5(StoreRecovery.java:285) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:308) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:283) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1867) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2580) ~[elasticsearch-7.5.1.jar:7.5.1]
	... 4 more
Caused by: java.io.IOException: Connection reset
	at sun.security.ssl.Alert.createSSLException(Alert.java:127) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:324) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:267) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:262) ~[?:?]
	at sun.security.ssl.SSLTransport.decode(SSLTransport.java:137) ~[?:?]
	at sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1463) ~[?:?]
	at sun.security.ssl.SSLSocketImpl.readApplicationRecord(SSLSocketImpl.java:1431) ~[?:?]
	at sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:1022) ~[?:?]
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:245) ~[?:?]
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:285) ~[?:?]
	at java.io.BufferedInputStream.read(BufferedInputStream.java:344) ~[?:?]
	at sun.net.www.MeteredStream.read(MeteredStream.java:134) ~[?:?]
	at java.io.FilterInputStream.read(FilterInputStream.java:133) ~[?:?]
	at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3505) ~[?:?]
	at com.google.api.client.http.javanet.NetHttpResponse$SizeValidatingInputStream.read(NetHttpResponse.java:164) ~[?:?]
	at java.io.FilterInputStream.read(FilterInputStream.java:107) ~[?:?]
	at com.google.api.client.util.ByteStreams.copy(ByteStreams.java:49) ~[?:?]
	at com.google.api.client.util.IOUtils.copy(IOUtils.java:87) ~[?:?]
	at com.google.api.client.util.IOUtils.copy(IOUtils.java:59) ~[?:?]
	at com.google.api.client.http.HttpResponse.download(HttpResponse.java:386) ~[?:?]
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.read(HttpStorageRpc.java:684) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:127) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:124) ~[?:?]
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105) ~[?:?]
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76) ~[?:?]
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel.read(BlobReadChannel.java:123) ~[?:?]
	at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.lambda$read$0(GoogleCloudStorageBlobStore.java:189) ~[?:?]
	at java.security.AccessController.doPrivileged(AccessController.java:554) ~[?:?]
	at org.elasticsearch.repositories.gcs.SocketAccess.doPrivilegedIOException(SocketAccess.java:44) ~[?:?]
	at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.read(GoogleCloudStorageBlobStore.java:189) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:65) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:109) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:103) ~[?:?]
	at org.elasticsearch.index.snapshots.blobstore.SlicedInputStream.read(SlicedInputStream.java:92) ~[elasticsearch-7.5.1.jar:7.5.1]
	at java.io.InputStream.read(InputStream.java:213) ~[?:?]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFile(FileRestoreContext.java:240) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFiles(FileRestoreContext.java:217) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restore(FileRestoreContext.java:180) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.restoreShard(BlobStoreRepository.java:1253) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.Repository.restoreShard(Repository.java:231) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:469) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromRepository$5(StoreRecovery.java:285) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:308) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:283) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1867) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2580) ~[elasticsearch-7.5.1.jar:7.5.1]
	... 4 more
	Suppressed: java.io.IOException: Broken pipe
		at sun.nio.ch.NioSocketImpl.implWrite(NioSocketImpl.java:421) ~[?:?]
		at sun.nio.ch.NioSocketImpl.write(NioSocketImpl.java:441) ~[?:?]
		at sun.nio.ch.NioSocketImpl$2.write(NioSocketImpl.java:825) ~[?:?]
		at java.net.Socket$SocketOutputStream.write(Socket.java:1007) ~[?:?]
		at sun.security.ssl.SSLSocketOutputRecord.encodeAlert(SSLSocketOutputRecord.java:82) ~[?:?]
		at sun.security.ssl.TransportContext.fatal(TransportContext.java:355) ~[?:?]
		at sun.security.ssl.TransportContext.fatal(TransportContext.java:267) ~[?:?]
		at sun.security.ssl.TransportContext.fatal(TransportContext.java:262) ~[?:?]
		at sun.security.ssl.SSLTransport.decode(SSLTransport.java:137) ~[?:?]
		at sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1463) ~[?:?]
		at sun.security.ssl.SSLSocketImpl.readApplicationRecord(SSLSocketImpl.java:1431) ~[?:?]
		at sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:1022) ~[?:?]
		at java.io.BufferedInputStream.fill(BufferedInputStream.java:245) ~[?:?]
		at java.io.BufferedInputStream.read1(BufferedInputStream.java:285) ~[?:?]
		at java.io.BufferedInputStream.read(BufferedInputStream.java:344) ~[?:?]
		at sun.net.www.MeteredStream.read(MeteredStream.java:134) ~[?:?]
		at java.io.FilterInputStream.read(FilterInputStream.java:133) ~[?:?]
		at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3505) ~[?:?]
		at com.google.api.client.http.javanet.NetHttpResponse$SizeValidatingInputStream.read(NetHttpResponse.java:164) ~[?:?]
		at java.io.FilterInputStream.read(FilterInputStream.java:107) ~[?:?]
		at com.google.api.client.util.ByteStreams.copy(ByteStreams.java:49) ~[?:?]
		at com.google.api.client.util.IOUtils.copy(IOUtils.java:87) ~[?:?]
		at com.google.api.client.util.IOUtils.copy(IOUtils.java:59) ~[?:?]
		at com.google.api.client.http.HttpResponse.download(HttpResponse.java:386) ~[?:?]
		at com.google.cloud.storage.spi.v1.HttpStorageRpc.read(HttpStorageRpc.java:684) ~[?:?]
		at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:127) ~[?:?]
		at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:124) ~[?:?]
		at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105) ~[?:?]
		at com.google.cloud.RetryHelper.run(RetryHelper.java:76) ~[?:?]
		at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50) ~[?:?]
		at com.google.cloud.storage.BlobReadChannel.read(BlobReadChannel.java:123) ~[?:?]
		at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.lambda$read$0(GoogleCloudStorageBlobStore.java:189) ~[?:?]
		at java.security.AccessController.doPrivileged(AccessController.java:554) ~[?:?]
		at org.elasticsearch.repositories.gcs.SocketAccess.doPrivilegedIOException(SocketAccess.java:44) ~[?:?]
		at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.read(GoogleCloudStorageBlobStore.java:189) ~[?:?]
		at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:65) ~[?:?]
		at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:109) ~[?:?]
		at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:103) ~[?:?]
		at org.elasticsearch.index.snapshots.blobstore.SlicedInputStream.read(SlicedInputStream.java:92) ~[elasticsearch-7.5.1.jar:7.5.1]
		at java.io.InputStream.read(InputStream.java:213) ~[?:?]
		at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFile(FileRestoreContext.java:240) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFiles(FileRestoreContext.java:217) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.repositories.blobstore.FileRestoreContext.restore(FileRestoreContext.java:180) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.repositories.blobstore.BlobStoreRepository.restoreShard(BlobStoreRepository.java:1253) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.repositories.Repository.restoreShard(Repository.java:231) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:469) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromRepository$5(StoreRecovery.java:285) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:308) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:283) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1867) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2580) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:703) ~[elasticsearch-7.5.1.jar:7.5.1]
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
		at java.lang.Thread.run(Thread.java:830) [?:?]
Caused by: java.io.IOException: Connection reset
	at sun.nio.ch.NioSocketImpl.implRead(NioSocketImpl.java:324) ~[?:?]
	at sun.nio.ch.NioSocketImpl.read(NioSocketImpl.java:351) ~[?:?]
	at sun.nio.ch.NioSocketImpl$1.read(NioSocketImpl.java:802) ~[?:?]
	at java.net.Socket$SocketInputStream.read(Socket.java:937) ~[?:?]
	at sun.security.ssl.SSLSocketInputRecord.read(SSLSocketInputRecord.java:450) ~[?:?]
	at sun.security.ssl.SSLSocketInputRecord.decodeInputRecord(SSLSocketInputRecord.java:237) ~[?:?]
	at sun.security.ssl.SSLSocketInputRecord.decode(SSLSocketInputRecord.java:190) ~[?:?]
	at sun.security.ssl.SSLTransport.decode(SSLTransport.java:108) ~[?:?]
	at sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1463) ~[?:?]
	at sun.security.ssl.SSLSocketImpl.readApplicationRecord(SSLSocketImpl.java:1431) ~[?:?]
	at sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:1022) ~[?:?]
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:245) ~[?:?]
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:285) ~[?:?]
	at java.io.BufferedInputStream.read(BufferedInputStream.java:344) ~[?:?]
	at sun.net.www.MeteredStream.read(MeteredStream.java:134) ~[?:?]
	at java.io.FilterInputStream.read(FilterInputStream.java:133) ~[?:?]
	at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3505) ~[?:?]
	at com.google.api.client.http.javanet.NetHttpResponse$SizeValidatingInputStream.read(NetHttpResponse.java:164) ~[?:?]
	at java.io.FilterInputStream.read(FilterInputStream.java:107) ~[?:?]
	at com.google.api.client.util.ByteStreams.copy(ByteStreams.java:49) ~[?:?]
	at com.google.api.client.util.IOUtils.copy(IOUtils.java:87) ~[?:?]
	at com.google.api.client.util.IOUtils.copy(IOUtils.java:59) ~[?:?]
	at com.google.api.client.http.HttpResponse.download(HttpResponse.java:386) ~[?:?]
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.read(HttpStorageRpc.java:684) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:127) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:124) ~[?:?]
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105) ~[?:?]
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76) ~[?:?]
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel.read(BlobReadChannel.java:123) ~[?:?]
	at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.lambda$read$0(GoogleCloudStorageBlobStore.java:189) ~[?:?]
	at java.security.AccessController.doPrivileged(AccessController.java:554) ~[?:?]
	at org.elasticsearch.repositories.gcs.SocketAccess.doPrivilegedIOException(SocketAccess.java:44) ~[?:?]
	at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.read(GoogleCloudStorageBlobStore.java:189) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:65) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:109) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:103) ~[?:?]
	at org.elasticsearch.index.snapshots.blobstore.SlicedInputStream.read(SlicedInputStream.java:92) ~[elasticsearch-7.5.1.jar:7.5.1]
	at java.io.InputStream.read(InputStream.java:213) ~[?:?]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFile(FileRestoreContext.java:240) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFiles(FileRestoreContext.java:217) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restore(FileRestoreContext.java:180) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.restoreShard(BlobStoreRepository.java:1253) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.Repository.restoreShard(Repository.java:231) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:469) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromRepository$5(StoreRecovery.java:285) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:308) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:283) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1867) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2580) ~[elasticsearch-7.5.1.jar:7.5.1]
	... 4 more
[2020-02-13T12:10:57,130][WARN ][o.e.i.c.IndicesClusterStateService] [elasticsearch-1] [elasticlogs-2018-05-13][0] marking and sending shard failed due to [failed recovery]
org.elasticsearch.indices.recovery.RecoveryFailedException: [elasticlogs-2018-05-13][0]: Recovery failed on {elasticsearch-1}{d2sAR621RziQbILiZMBvfg}{cgzSxj1tSjawYThZfSEXpQ}{10.10.9.72}{10.10.9.72:9300}{dim}{xpack.installed=true}
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2584) [elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:703) [elasticsearch-7.5.1.jar:7.5.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:830) [?:?]
Caused by: org.elasticsearch.index.shard.IndexShardRecoveryException: failed recovery
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:353) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:283) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1867) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2580) ~[elasticsearch-7.5.1.jar:7.5.1]
	... 4 more
Caused by: org.elasticsearch.index.snapshots.IndexShardRestoreFailedException: restore failed
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:480) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromRepository$5(StoreRecovery.java:285) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:308) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:283) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1867) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2580) ~[elasticsearch-7.5.1.jar:7.5.1]
	... 4 more
Caused by: org.elasticsearch.index.snapshots.IndexShardRestoreFailedException: failed to restore snapshot [snapshot_80gb/fioJzVMyTcKPCRdfN5lZ7w]
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.restoreShard(BlobStoreRepository.java:1255) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.Repository.restoreShard(Repository.java:231) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:469) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromRepository$5(StoreRecovery.java:285) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:308) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:283) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1867) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2580) ~[elasticsearch-7.5.1.jar:7.5.1]
	... 4 more
Caused by: com.google.cloud.storage.StorageException: Connection reset
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:227) ~[?:?]
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.read(HttpStorageRpc.java:690) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:127) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:124) ~[?:?]
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105) ~[?:?]
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76) ~[?:?]
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel.read(BlobReadChannel.java:123) ~[?:?]
	at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.lambda$read$0(GoogleCloudStorageBlobStore.java:189) ~[?:?]
	at java.security.AccessController.doPrivileged(AccessController.java:554) ~[?:?]
	at org.elasticsearch.repositories.gcs.SocketAccess.doPrivilegedIOException(SocketAccess.java:44) ~[?:?]
	at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.read(GoogleCloudStorageBlobStore.java:189) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:65) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:109) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:103) ~[?:?]
	at org.elasticsearch.index.snapshots.blobstore.SlicedInputStream.read(SlicedInputStream.java:92) ~[elasticsearch-7.5.1.jar:7.5.1]
	at java.io.InputStream.read(InputStream.java:213) ~[?:?]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFile(FileRestoreContext.java:240) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFiles(FileRestoreContext.java:217) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restore(FileRestoreContext.java:180) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.restoreShard(BlobStoreRepository.java:1253) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.Repository.restoreShard(Repository.java:231) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:469) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromRepository$5(StoreRecovery.java:285) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:308) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:283) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1867) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2580) ~[elasticsearch-7.5.1.jar:7.5.1]
	... 4 more
Caused by: javax.net.ssl.SSLException: Connection reset
	at sun.security.ssl.Alert.createSSLException(Alert.java:127) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:324) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:267) ~[?:?]
	at sun.security.ssl.TransportContext.fatal(TransportContext.java:262) ~[?:?]
	at sun.security.ssl.SSLTransport.decode(SSLTransport.java:137) ~[?:?]
	at sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1463) ~[?:?]
	at sun.security.ssl.SSLSocketImpl.readApplicationRecord(SSLSocketImpl.java:1431) ~[?:?]
	at sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:1022) ~[?:?]
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:245) ~[?:?]
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:285) ~[?:?]
	at java.io.BufferedInputStream.read(BufferedInputStream.java:344) ~[?:?]
	at sun.net.www.MeteredStream.read(MeteredStream.java:134) ~[?:?]
	at java.io.FilterInputStream.read(FilterInputStream.java:133) ~[?:?]
	at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3505) ~[?:?]
	at com.google.api.client.http.javanet.NetHttpResponse$SizeValidatingInputStream.read(NetHttpResponse.java:164) ~[?:?]
	at java.io.FilterInputStream.read(FilterInputStream.java:107) ~[?:?]
	at com.google.api.client.util.ByteStreams.copy(ByteStreams.java:49) ~[?:?]
	at com.google.api.client.util.IOUtils.copy(IOUtils.java:87) ~[?:?]
	at com.google.api.client.util.IOUtils.copy(IOUtils.java:59) ~[?:?]
	at com.google.api.client.http.HttpResponse.download(HttpResponse.java:386) ~[?:?]
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.read(HttpStorageRpc.java:684) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:127) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:124) ~[?:?]
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105) ~[?:?]
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76) ~[?:?]
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel.read(BlobReadChannel.java:123) ~[?:?]
	at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.lambda$read$0(GoogleCloudStorageBlobStore.java:189) ~[?:?]
	at java.security.AccessController.doPrivileged(AccessController.java:554) ~[?:?]
	at org.elasticsearch.repositories.gcs.SocketAccess.doPrivilegedIOException(SocketAccess.java:44) ~[?:?]
	at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.read(GoogleCloudStorageBlobStore.java:189) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:65) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:109) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:103) ~[?:?]
	at org.elasticsearch.index.snapshots.blobstore.SlicedInputStream.read(SlicedInputStream.java:92) ~[elasticsearch-7.5.1.jar:7.5.1]
	at java.io.InputStream.read(InputStream.java:213) ~[?:?]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFile(FileRestoreContext.java:240) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFiles(FileRestoreContext.java:217) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restore(FileRestoreContext.java:180) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.restoreShard(BlobStoreRepository.java:1253) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.Repository.restoreShard(Repository.java:231) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:469) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromRepository$5(StoreRecovery.java:285) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:308) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:283) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1867) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2580) ~[elasticsearch-7.5.1.jar:7.5.1]
	... 4 more
	Suppressed: java.net.SocketException: Broken pipe
		at sun.nio.ch.NioSocketImpl.implWrite(NioSocketImpl.java:421) ~[?:?]
		at sun.nio.ch.NioSocketImpl.write(NioSocketImpl.java:441) ~[?:?]
		at sun.nio.ch.NioSocketImpl$2.write(NioSocketImpl.java:825) ~[?:?]
		at java.net.Socket$SocketOutputStream.write(Socket.java:1007) ~[?:?]
		at sun.security.ssl.SSLSocketOutputRecord.encodeAlert(SSLSocketOutputRecord.java:82) ~[?:?]
		at sun.security.ssl.TransportContext.fatal(TransportContext.java:355) ~[?:?]
		at sun.security.ssl.TransportContext.fatal(TransportContext.java:267) ~[?:?]
		at sun.security.ssl.TransportContext.fatal(TransportContext.java:262) ~[?:?]
		at sun.security.ssl.SSLTransport.decode(SSLTransport.java:137) ~[?:?]
		at sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1463) ~[?:?]
		at sun.security.ssl.SSLSocketImpl.readApplicationRecord(SSLSocketImpl.java:1431) ~[?:?]
		at sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:1022) ~[?:?]
		at java.io.BufferedInputStream.fill(BufferedInputStream.java:245) ~[?:?]
		at java.io.BufferedInputStream.read1(BufferedInputStream.java:285) ~[?:?]
		at java.io.BufferedInputStream.read(BufferedInputStream.java:344) ~[?:?]
		at sun.net.www.MeteredStream.read(MeteredStream.java:134) ~[?:?]
		at java.io.FilterInputStream.read(FilterInputStream.java:133) ~[?:?]
		at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3505) ~[?:?]
		at com.google.api.client.http.javanet.NetHttpResponse$SizeValidatingInputStream.read(NetHttpResponse.java:164) ~[?:?]
		at java.io.FilterInputStream.read(FilterInputStream.java:107) ~[?:?]
		at com.google.api.client.util.ByteStreams.copy(ByteStreams.java:49) ~[?:?]
		at com.google.api.client.util.IOUtils.copy(IOUtils.java:87) ~[?:?]
		at com.google.api.client.util.IOUtils.copy(IOUtils.java:59) ~[?:?]
		at com.google.api.client.http.HttpResponse.download(HttpResponse.java:386) ~[?:?]
		at com.google.cloud.storage.spi.v1.HttpStorageRpc.read(HttpStorageRpc.java:684) ~[?:?]
		at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:127) ~[?:?]
		at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:124) ~[?:?]
		at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105) ~[?:?]
		at com.google.cloud.RetryHelper.run(RetryHelper.java:76) ~[?:?]
		at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50) ~[?:?]
		at com.google.cloud.storage.BlobReadChannel.read(BlobReadChannel.java:123) ~[?:?]
		at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.lambda$read$0(GoogleCloudStorageBlobStore.java:189) ~[?:?]
		at java.security.AccessController.doPrivileged(AccessController.java:554) ~[?:?]
		at org.elasticsearch.repositories.gcs.SocketAccess.doPrivilegedIOException(SocketAccess.java:44) ~[?:?]
		at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.read(GoogleCloudStorageBlobStore.java:189) ~[?:?]
		at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:65) ~[?:?]
		at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:109) ~[?:?]
		at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:103) ~[?:?]
		at org.elasticsearch.index.snapshots.blobstore.SlicedInputStream.read(SlicedInputStream.java:92) ~[elasticsearch-7.5.1.jar:7.5.1]
		at java.io.InputStream.read(InputStream.java:213) ~[?:?]
		at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFile(FileRestoreContext.java:240) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFiles(FileRestoreContext.java:217) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.repositories.blobstore.FileRestoreContext.restore(FileRestoreContext.java:180) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.repositories.blobstore.BlobStoreRepository.restoreShard(BlobStoreRepository.java:1253) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.repositories.Repository.restoreShard(Repository.java:231) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:469) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromRepository$5(StoreRecovery.java:285) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:308) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:283) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1867) ~[elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2580) [elasticsearch-7.5.1.jar:7.5.1]
		at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:703) [elasticsearch-7.5.1.jar:7.5.1]
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
		at java.lang.Thread.run(Thread.java:830) [?:?]
Caused by: java.net.SocketException: Connection reset
	at sun.nio.ch.NioSocketImpl.implRead(NioSocketImpl.java:324) ~[?:?]
	at sun.nio.ch.NioSocketImpl.read(NioSocketImpl.java:351) ~[?:?]
	at sun.nio.ch.NioSocketImpl$1.read(NioSocketImpl.java:802) ~[?:?]
	at java.net.Socket$SocketInputStream.read(Socket.java:937) ~[?:?]
	at sun.security.ssl.SSLSocketInputRecord.read(SSLSocketInputRecord.java:450) ~[?:?]
	at sun.security.ssl.SSLSocketInputRecord.decodeInputRecord(SSLSocketInputRecord.java:237) ~[?:?]
	at sun.security.ssl.SSLSocketInputRecord.decode(SSLSocketInputRecord.java:190) ~[?:?]
	at sun.security.ssl.SSLTransport.decode(SSLTransport.java:108) ~[?:?]
	at sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1463) ~[?:?]
	at sun.security.ssl.SSLSocketImpl.readApplicationRecord(SSLSocketImpl.java:1431) ~[?:?]
	at sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:1022) ~[?:?]
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:245) ~[?:?]
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:285) ~[?:?]
	at java.io.BufferedInputStream.read(BufferedInputStream.java:344) ~[?:?]
	at sun.net.www.MeteredStream.read(MeteredStream.java:134) ~[?:?]
	at java.io.FilterInputStream.read(FilterInputStream.java:133) ~[?:?]
	at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3505) ~[?:?]
	at com.google.api.client.http.javanet.NetHttpResponse$SizeValidatingInputStream.read(NetHttpResponse.java:164) ~[?:?]
	at java.io.FilterInputStream.read(FilterInputStream.java:107) ~[?:?]
	at com.google.api.client.util.ByteStreams.copy(ByteStreams.java:49) ~[?:?]
	at com.google.api.client.util.IOUtils.copy(IOUtils.java:87) ~[?:?]
	at com.google.api.client.util.IOUtils.copy(IOUtils.java:59) ~[?:?]
	at com.google.api.client.http.HttpResponse.download(HttpResponse.java:386) ~[?:?]
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.read(HttpStorageRpc.java:684) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:127) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel$1.call(BlobReadChannel.java:124) ~[?:?]
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:105) ~[?:?]
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76) ~[?:?]
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50) ~[?:?]
	at com.google.cloud.storage.BlobReadChannel.read(BlobReadChannel.java:123) ~[?:?]
	at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.lambda$read$0(GoogleCloudStorageBlobStore.java:189) ~[?:?]
	at java.security.AccessController.doPrivileged(AccessController.java:554) ~[?:?]
	at org.elasticsearch.repositories.gcs.SocketAccess.doPrivilegedIOException(SocketAccess.java:44) ~[?:?]
	at org.elasticsearch.repositories.gcs.GoogleCloudStorageBlobStore$1.read(GoogleCloudStorageBlobStore.java:189) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:65) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:109) ~[?:?]
	at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:103) ~[?:?]
	at org.elasticsearch.index.snapshots.blobstore.SlicedInputStream.read(SlicedInputStream.java:92) ~[elasticsearch-7.5.1.jar:7.5.1]
	at java.io.InputStream.read(InputStream.java:213) ~[?:?]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFile(FileRestoreContext.java:240) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restoreFiles(FileRestoreContext.java:217) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.FileRestoreContext.restore(FileRestoreContext.java:180) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.restoreShard(BlobStoreRepository.java:1253) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.repositories.Repository.restoreShard(Repository.java:231) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.restore(StoreRecovery.java:469) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.lambda$recoverFromRepository$5(StoreRecovery.java:285) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.executeRecovery(StoreRecovery.java:308) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.StoreRecovery.recoverFromRepository(StoreRecovery.java:283) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.restoreFromRepository(IndexShard.java:1867) ~[elasticsearch-7.5.1.jar:7.5.1]
	at org.elasticsearch.index.shard.IndexShard.lambda$startRecovery$17(IndexShard.java:2580) ~[elasticsearch-7.5.1.jar:7.5.1]
	... 4 more
```
</p>
</details>

l.e. I've seen `java.io.IOException: Connection reset` and `javax.net.ssl.SSLException: Connection reset`, [neither of which seem to be retryable](https://github.com/googleapis/java-core/blob/573d42d704413ee10dc2e600e8891575508ab7b7/google-cloud-core/src/main/java/com/google/cloud/BaseServiceException.java#L254-L264)

Relates #46589, which works around the same issue in the AWS S3 SDK.

Quoting https://github.com/googleapis/java-storage/issues/26#issuecomment-575144315:

> network issues [...] are out of the scope of its responsibility

This seems to ignore the fact that connection drops are often transient and resuming the download is often the right way to proceed. Still, at least it's a clear statement of intent, and indicates that maybe we need to implement our own resume functionality to work around this for now.

--- a/plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java
+++ b/plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java
@@ -21,7 +21,6 @@ package org.elasticsearch.repositories.gcs;
 
 import com.google.api.gax.paging.Page;
 import com.google.cloud.BatchResult;
-import com.google.cloud.ReadChannel;
 import com.google.cloud.storage.Blob;
 import com.google.cloud.storage.BlobId;
 import com.google.cloud.storage.BlobInfo;
@@ -34,7 +33,6 @@ import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.logging.log4j.message.ParameterizedMessage;
 import org.elasticsearch.ExceptionsHelper;
-import org.elasticsearch.common.SuppressForbidden;
 import org.elasticsearch.common.blobstore.BlobContainer;
 import org.elasticsearch.common.blobstore.BlobMetaData;
 import org.elasticsearch.common.blobstore.BlobPath;
@@ -47,11 +45,8 @@ import org.elasticsearch.core.internal.io.Streams;
 
 import java.io.IOException;
 import java.io.InputStream;
-import java.nio.ByteBuffer;
 import java.nio.channels.Channels;
-import java.nio.channels.ReadableByteChannel;
 import java.nio.file.FileAlreadyExistsException;
-import java.nio.file.NoSuchFileException;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
@@ -176,32 +171,7 @@ class GoogleCloudStorageBlobStore implements BlobStore {
      * @return the InputStream used to read the blob's content
      */
     InputStream readBlob(String blobName) throws IOException {
-        final BlobId blobId = BlobId.of(bucketName, blobName);
-        final ReadChannel readChannel = SocketAccess.doPrivilegedIOException(() -> client().reader(blobId));
-        return Channels.newInputStream(new ReadableByteChannel() {
-            @SuppressForbidden(reason = "Channel is based of a socket not a file")
-            @Override
-            public int read(ByteBuffer dst) throws IOException {
-                try {
-                    return SocketAccess.doPrivilegedIOException(() -> readChannel.read(dst));
-                } catch (StorageException e) {
-                    if (e.getCode() == HTTP_NOT_FOUND) {
-                        throw new NoSuchFileException("Blob [" + blobName + "] does not exist");
-                    }
-                    throw e;
-                }
-            }
-
-            @Override
-            public boolean isOpen() {
-                return readChannel.isOpen();
-            }
-
-            @Override
-            public void close() throws IOException {
-                SocketAccess.doPrivilegedVoidIOException(readChannel::close);
-            }
-        });
+        return new GoogleCloudStorageRetryingInputStream(client(), BlobId.of(bucketName, blobName));
     }
 
     /**

--- a/plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java
+++ b/plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java
@@ -1,0 +1,181 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.repositories.gcs;
+
+import com.google.cloud.ReadChannel;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.common.SuppressForbidden;
+import org.elasticsearch.core.internal.io.IOUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.NoSuchFileException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.net.HttpURLConnection.HTTP_NOT_FOUND;
+
+/**
+ * Wrapper around reads from GCS that will retry blob downloads that fail part-way through, resuming from where the failure occurred.
+ * This should be handled by the SDK but it isn't today. This should be revisited in the future (e.g. before removing
+ * the {@link org.elasticsearch.Version#V_7_0_0} version constant) and removed if the SDK handles retries itself in the future.
+ */
+class GoogleCloudStorageRetryingInputStream extends InputStream {
+
+    private static final Logger logger = LogManager.getLogger(GoogleCloudStorageRetryingInputStream.class);
+
+    static final int MAX_SUPPRESSED_EXCEPTIONS = 10;
+
+    private final Storage client;
+
+    private final BlobId blobId;
+
+    private final int maxRetries;
+
+    private InputStream currentStream;
+    private int attempt = 1;
+    private List<StorageException> failures = new ArrayList<>(MAX_SUPPRESSED_EXCEPTIONS);
+    private long currentOffset;
+    private boolean closed;
+
+    GoogleCloudStorageRetryingInputStream(Storage client, BlobId blobId) throws IOException {
+        this.client = client;
+        this.blobId = blobId;
+        this.maxRetries = client.getOptions().getRetrySettings().getMaxAttempts() + 1;
+        currentStream = openStream();
+    }
+
+    private InputStream openStream() throws IOException {
+        try {
+            final ReadChannel readChannel = SocketAccess.doPrivilegedIOException(() -> client.reader(blobId));
+            if (currentOffset > 0L) {
+                readChannel.seek(currentOffset);
+            }
+            return Channels.newInputStream(new ReadableByteChannel() {
+                @SuppressForbidden(reason = "Channel is based of a socket not a file")
+                @Override
+                public int read(ByteBuffer dst) throws IOException {
+                    try {
+                        return SocketAccess.doPrivilegedIOException(() -> readChannel.read(dst));
+                    } catch (StorageException e) {
+                        if (e.getCode() == HTTP_NOT_FOUND) {
+                            throw new NoSuchFileException("Blob [" + blobId.getName() + "] does not exist");
+                        }
+                        throw e;
+                    }
+                }
+
+                @Override
+                public boolean isOpen() {
+                    return readChannel.isOpen();
+                }
+
+                @Override
+                public void close() throws IOException {
+                    SocketAccess.doPrivilegedVoidIOException(readChannel::close);
+                }
+            });
+        } catch (StorageException e) {
+            throw addSuppressedExceptions(e);
+        }
+    }
+
+    @Override
+    public int read() throws IOException {
+        ensureOpen();
+        while (true) {
+            try {
+                final int result = currentStream.read();
+                currentOffset += 1;
+                return result;
+            } catch (StorageException e) {
+                reopenStreamOrFail(e);
+            }
+        }
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        ensureOpen();
+        while (true) {
+            try {
+                final int bytesRead = currentStream.read(b, off, len);
+                if (bytesRead == -1) {
+                    return -1;
+                }
+                currentOffset += bytesRead;
+                return bytesRead;
+            } catch (StorageException e) {
+                reopenStreamOrFail(e);
+            }
+        }
+    }
+
+    private void ensureOpen() {
+        if (closed) {
+            assert false : "using GoogleCloudStorageRetryingInputStream after close";
+            throw new IllegalStateException("using GoogleCloudStorageRetryingInputStream after close");
+        }
+    }
+
+    private void reopenStreamOrFail(StorageException e) throws IOException {
+        if (attempt >= maxRetries) {
+            throw addSuppressedExceptions(e);
+        }
+        logger.debug(new ParameterizedMessage("failed reading [{}] at offset [{}], attempt [{}] of [{}], retrying",
+            blobId, currentOffset, attempt, MAX_SUPPRESSED_EXCEPTIONS), e);
+        attempt += 1;
+        if (failures.size() < MAX_SUPPRESSED_EXCEPTIONS) {
+            failures.add(e);
+        }
+        IOUtils.closeWhileHandlingException(currentStream);
+        currentStream = openStream();
+    }
+
+    @Override
+    public void close() throws IOException {
+        currentStream.close();
+        closed = true;
+    }
+
+    @Override
+    public long skip(long n) {
+        throw new UnsupportedOperationException("GoogleCloudStorageRetryingInputStream does not support seeking");
+    }
+
+    @Override
+    public void reset() {
+        throw new UnsupportedOperationException("GoogleCloudStorageRetryingInputStream does not support seeking");
+    }
+
+    private <T extends Exception> T addSuppressedExceptions(T e) {
+        for (StorageException failure : failures) {
+            e.addSuppressed(failure);
+        }
+        return e;
+    }
+}

--- a/plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobContainerRetriesTests.java
+++ b/plugins/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobContainerRetriesTests.java
@@ -200,6 +200,35 @@ public class GoogleCloudStorageBlobContainerRetriesTests extends ESTestCase {
         }
     }
 
+    public void testReadLargeBlobWithRetries() throws Exception {
+        final int maxRetries = randomIntBetween(2, 10);
+        final CountDown countDown = new CountDown(maxRetries);
+
+        // SDK reads in 2 MB chunks so we use twice that to simulate 2 chunks
+        final byte[] bytes = randomBytes(1 << 22);
+        httpServer.createContext("/download/storage/v1/b/bucket/o/large_blob_retries", exchange -> {
+            Streams.readFully(exchange.getRequestBody());
+            exchange.getResponseHeaders().add("Content-Type", "application/octet-stream");
+            final String[] range = exchange.getRequestHeaders().get("Range").get(0).substring("bytes=".length()).split("-");
+            final int offset = Integer.parseInt(range[0]);
+            final int end = Integer.parseInt(range[1]);
+            final byte[] chunk = Arrays.copyOfRange(bytes, offset, Math.min(end + 1, bytes.length));
+            exchange.sendResponseHeaders(RestStatus.OK.getStatus(), chunk.length);
+            if (randomBoolean() && countDown.countDown() == false) {
+                exchange.getResponseBody().write(chunk, 0, chunk.length - 1);
+                exchange.close();
+                return;
+            }
+            exchange.getResponseBody().write(chunk);
+            exchange.close();
+        });
+
+        final BlobContainer blobContainer = createBlobContainer(maxRetries, null);
+        try (InputStream inputStream = blobContainer.readBlob("large_blob_retries")) {
+            assertArrayEquals(bytes, BytesReference.toBytes(Streams.readFully(inputStream)));
+        }
+    }
+
     public void testReadBlobWithReadTimeouts() {
         final int maxRetries = randomIntBetween(1, 3);
         final BlobContainer blobContainer = createBlobContainer(maxRetries, TimeValue.timeValueMillis(between(100, 200)));

